### PR TITLE
resource resolver extension for iOS

### PR DESCRIPTION
**Pull request by nesalang**
allows users to use their own async downloading process

### CODE REVIEW

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOResourceResolvers.mm [hunk 1]
```diff
@@ -10,25 +10,43 @@
...
 {
     self->_resolvers[scheme] = resolver;
+
+    _resolversIFMap[scheme] = [NSNumber numberWithInt:ACODefaultIF];
+
+    if([resolver respondsToSelector:@selector(resolveImageResource:)]) {
+        _resolversIFMap[scheme] = [NSNumber numberWithInt:ACOImageIF];
+    }
+    // only one IF per scheme is supported and ACRImageViewIF will be chosen
+    // when both respod
```
Comment (paulcam206) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOResourceResolvers.mm]: >d [](start = 22, length = 1)

nitpick - typo :)

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRImageRenderer.mm [hunk 2]
```diff
@@ -85,54 +56,38 @@ - (UIView *)render:(UIView<ACRIContentHoldingView> *)viewGroup
...
+        ACRUIImageView *acrImageView = [[ACRUIImageView alloc] initWithFrame:CGRectMake(0, 0, cgsize.width, cgsize.height)];
+        acrImageView.image = img;
+        if(imgElem->GetImageStyle() == ImageStyle::Person) {
+            acrImageView.isPersonStyle = YES;
+            [acrImageView setNeedsLayout];
+        }
+        view = acrImageView;
+        [self configUpdateForUIImageView:acoElem config:acoConfig image:img imageView:view];
+    } else
+    {
```
Comment (paulcam206) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRImageRenderer.mm]: does the `{` go on the same line as the else here? inconsistent.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm [hunk 3]
```diff
@@ -129,7 +125,22 @@ - (UIView *)render:(UIView<ACRIContentHoldingView> *)viewGroup
...
 
+    [self configUpdateForMediaView:heightToWidthRatio viewGroup:viewGroup contentHoldingView:contentholdingview imageView:view];
+
     return view;
 }
 
+//- (void)configUpdateForMediaView:(ACOBaseCardElement *)acoElem config:(ACOHostConfig *)acoConfig image:(UIImage *)image imageView:(UIImageView *)imageView
+- (void)configUpdateForMediaView:(CGFloat)heightToWidthRatio viewGroup:(UIView *)viewGroup contentHoldingView:(UIView *)contentholdingview imageView:(UIImageView *)view
+{
+    //heightToWidthRatio = img.size.height / img.size.width;
```
Comment (paulcam206) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm]: delete commented code

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRView.mm [hunk 4]
```diff
@@ -100,8 +103,15 @@ - (UIView *)render
...
         img = _imageViewMap[key];
-        imgView = [[ACRUIImageView alloc] initWithImage:img];
-        if(img) {
+        if(img)
+        {
+            imgView = [[ACRUIImageView alloc] initWithImage:img];
+        } else
+        {
+            imgView = [self getImageView:@"backgroundImage"];
+        }
```
Comment (paulcam206) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRView.mm]: inconsistent curly braces

File: source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/main.m [hunk 5]
```diff
@@ -11,6 +11,6 @@
 
 int main(int argc, char * argv[]) {
     @autoreleasepool {
-        return UIApplicationMain(argc, argv, nil, NSStringFromClass([AppDelegate class]));
+	        return UIApplicationMain(argc, argv, nil, NSStringFromClass([AppDelegate class]));
```
Comment (paulcam206) [inline source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/main.m]: >	 [](start = 0, length = 1)

remove tab

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRView.mm [hunk 6]
```diff
@@ -283,7 +315,18 @@ - (void)addTasksToConcurrentQueue:(std::vector<std::shared_ptr<BaseCardElement>>
 - (void)loadImagesForActionsAndCheckIfAllActionsHaveIconImages:(std::vector<std::shared_ptr<BaseActionElement>> const &)actions hostconfig:(ACOHostConfig *)hostConfig;
 {
     for(auto &action : actions){
-        if(action->GetIconUrl().size()) {
+        if(!action->GetIconUrl().empty()) {
+            ObserverActionBlock observerAction =
```
Comment (tnev) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRView.mm]: Is it intentional that this observerAction is not being used? The render methods for images above use their ObserverActionBlocks for subsequent -loadImageAccordingToResourceResolverIF calls, but this is just calling -loadImage.

**Comment by tnev** [comment]
@nesalang Thanks for pointing me to this branch! This definitely solves the issue I was facing of images loading on the main thread. I created a resource resolver much like the one that you have in the sample project, but I had to implement both ResourceResolver methods in order for it to work.

The only thing that seems to be needed for this to be a complete solution for using Cards in a UITableViewCell is public exposure of when ACRView's -observeValueForKeyPath is done so that you can tell the table to update its cell heights. Currently, if a cell is loaded before a height-determining image in that cell has loaded the cell will not layout correctly. Basically we need something sort of like the -didLoadElements method of ACRActionDelegate but to be called right after -configUpdateForUIImageView in ACRView's -observeValueForKeyPath.

Thanks again for the help!

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRView.mm (hunk 6 above)
Comment (nesalang) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRView.mm]: yes, this is intentional. the support has not been added for actions, yet. 

---
In reply to: [237271207](https://github.com/Microsoft/AdaptiveCards/pull/2169#discussion_r237271207) [](ancestors = 237271207)

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOResourceResolvers.mm (hunk 1 above)
Comment (nesalang) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOResourceResolvers.mm]: thanks!

---
In reply to: [236838055](https://github.com/Microsoft/AdaptiveCards/pull/2169#discussion_r236838055) [](ancestors = 236838055)

**Comment by nesalang** [comment]
thanks for the feedback; i will get back to you about that.

---
In reply to: [442641738](https://github.com/Microsoft/AdaptiveCards/pull/2169#issuecomment-442641738) [](ancestors = 442641738)

**Comment by tnev** [comment]
@nesalang Correct, the UITableViewCell won't call -layoutSubviews after the image view is resolved. I could rig something up so that the ResourceResolver would notify the table after it's done resolving the image views, but it seems like since you've already set up the observer in ACRView to listen for that it might be an opportune time to also tell the table to update its cell heights.

**Comment by nesalang** [comment]
okay, try the latest of the branch.

---
In reply to: [442933600](https://github.com/Microsoft/AdaptiveCards/pull/2169#issuecomment-442933600) [](ancestors = 442933600)

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm [hunk 7]
```diff
@@ -45,70 +44,81 @@ - (UIView *)render:(UIView<ACRIContentHoldingView> *)viewGroup
...
     NSString *piikey = [NSString stringWithCString:[acoConfig getHostConfig]->GetMedia().playButton.c_str() encoding:[NSString defaultCStringEncoding]];
     UIImage *playIconImage = imageViewMap[piikey];
     UIImageView *playIconImageView = nil;
     BOOL drawDefaultPlayIcon = YES;
 
+    if(!playIconImage) {
+        playIconImageView  = [rootView getImageView:@"playIconImage"];
+    }
+
     if(playIconImage) {
```
Comment (paulcam206) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaRenderer.mm]: if/else?

**Comment by paulcam206** [comment]
> */

why the commented-out code here?

---
Refers to: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRMediaTarget.mm:138 in 322c042. [](commit_id = 322c042bf838f44d070f051203d640918874d7d1, deletion_comment = True)

**Comment by nesalang** [comment]
closing this PR; please use [this branch](https://github.com/Microsoft/AdaptiveCards/pull/2197).

**Pull request closed without merging**